cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -233,6 +235,184 @@ struct DivideChecked {
   }
 };
 
+struct Power {
+  template <typename T, typename Arg0, typename Arg1>
+  static enable_if_signed_integer<T> Call(KernelContext* ctx, Arg0 left, Arg1 right) {
+    if (right < 0) {
+      ctx->SetStatus(
+          Status::Invalid("integers to negative integer powers are not allowed"));
+    }
+    if (left == 0 && right != 0) {
+      return 0;
+    }
+    Arg0 result = 1;
+    for (Arg1 i = 0; i < right; i++) {
+      MultiplyWithOverflow(result, left, &result);
+    }
+    return result;
+  }
+
+  template <typename T, typename Arg0, typename Arg1>
+  static enable_if_unsigned_integer<T> Call(KernelContext* ctx, Arg0 left, Arg1 right) {
+    if (left == 0 && right != 0) {
+      return 0;
+    }
+    Arg0 result = 1;
+    for (Arg1 i = 0; i < right; i++) {
+      MultiplyWithOverflow(result, left, &result);
+    }
+    return result;
+  }
+
+  template <typename T, typename Arg0, typename Arg1>
+  static enable_if_floating_point<T> Call(KernelContext* ctx, Arg0 left, Arg1 right) {
+    if (std::isnan(left) || std::isnan(right)) {
+      return NAN;
+    }
+    if (left == 0 && right < 0) {
+      ctx->SetStatus(Status::Invalid("divide by zero"));
+    }
+    return pow(left, right);
+  }
+};
+
+struct PowerChecked {
```

Comment:
I see many similar codes in this change. Should refine to keep code concise.

Reply:
I've abstracted a bit, perhaps more is possible.